File: code/10-state-space-search/c/a_star_grid.c
/* lecture-notes/code/lecture10/c/a_star_grid.c와 같은 정책(4-neighbor
 * unit-cost grid, Manhattan heuristic, stale-g-snapshot 검사, permanent
 * CLOSED). */
#include "state_space_search.h"
#include <limits.h>
#include <stdlib.h>

// snippet:a-star:start
typedef struct { size_t cell; int64_t g; int64_t f; } entry;

static int less(entry a, entry b) {
    if (a.f != b.f) return a.f < b.f;
    if (a.g != b.g) return a.g < b.g;
    return a.cell < b.cell;
}
static void push(entry *h, size_t *n, entry x) {
    size_t i = (*n)++;
    while (i > 0U) {
        size_t p = (i - 1U) / 2U;
        if (!less(x, h[p])) break;
        h[i] = h[p]; i = p;
    }
    h[i] = x;
}
static entry pop(entry *h, size_t *n) {
    entry top = h[0], x = h[--(*n)];
    size_t i = 0U;
    while (2U * i + 1U < *n) {
        size_t c = 2U * i + 1U;
        if (c + 1U < *n && less(h[c + 1U], h[c])) c++;
        if (!less(h[c], x)) break;
        h[i] = h[c]; i = c;
    }
    if (*n != 0U) h[i] = x;
    return top;
}
static int64_t heuristic(size_t cell, size_t goal, size_t cols, bool zero) {
    size_t r1 = cell / cols, c1 = cell % cols;
    size_t r2 = goal / cols, c2 = goal % cols;
    if (zero) return 0;
    return (int64_t)(r1 > r2 ? r1-r2 : r2-r1) +
           (int64_t)(c1 > c2 ? c1-c2 : c2-c1);
}

ss_status ss_astar_grid(const unsigned char *blocked, size_t rows, size_t cols,
                        size_t start, size_t goal, bool zero_heuristic,
                        int64_t *cost, ss_metrics *metrics) {
    size_t cells, heap_cap, heap_size = 0U, i;
    int64_t *g;
    unsigned char *closed;
    entry *heap;
    static const int dr[4] = {-1, 0, 0, 1};
    static const int dc[4] = {0, -1, 1, 0};
    ss_metrics local = {0U,0U,0U,0U};
    if (blocked == NULL || cost == NULL || metrics == NULL || rows == 0U ||
        cols == 0U || rows > SIZE_MAX / cols) return SS_INVALID;
    cells = rows * cols;
    if (start >= cells || goal >= cells || blocked[start] || blocked[goal] ||
        cells > (SIZE_MAX - 1U) / 4U) return SS_INVALID;
    heap_cap = 4U * cells + 1U;
    g = malloc(cells * sizeof(*g));
    closed = calloc(cells, sizeof(*closed));
    heap = malloc(heap_cap * sizeof(*heap));
    if (g == NULL || closed == NULL || heap == NULL) {
        free(g); free(closed); free(heap); return SS_NO_MEMORY;
    }
    for (i = 0; i < cells; i++) g[i] = INT64_MAX;
    g[start] = 0;
    push(heap,&heap_size,(entry){start,0,heuristic(start,goal,cols,zero_heuristic)});
    *cost = INT64_MAX;
    while (heap_size != 0U) {
        entry e = pop(heap,&heap_size);
        size_t r = e.cell / cols, c = e.cell % cols;
        int d;
        if (e.g != g[e.cell]) continue;
        if (closed[e.cell] != 0U) continue;
        closed[e.cell] = 1U;
        local.expanded++;
        if (heap_size > local.max_frontier) local.max_frontier = heap_size;
        if (e.cell == goal) { *cost = e.g; break; }
        for (d = 0; d < 4; d++) {
            int64_t nr = (int64_t)r + dr[d], nc = (int64_t)c + dc[d];
            size_t v;
            int64_t ng;
            if (nr < 0 || nc < 0 || nr >= (int64_t)rows || nc >= (int64_t)cols) continue;
            v = (size_t)nr * cols + (size_t)nc;
            if (blocked[v] || closed[v] != 0U) continue;
            ng = e.g + 1;
            if (ng < g[v]) {
                g[v] = ng;
                push(heap,&heap_size,(entry){v,ng,ng+heuristic(v,goal,cols,zero_heuristic)});
            }
        }
    }
    free(g); free(closed); free(heap); *metrics = local;
    return *cost == INT64_MAX ? SS_NO_SOLUTION : SS_OK;
}
// snippet:a-star:end

File: code/10-state-space-search/java/AStarGrid.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.PriorityQueue;

/** lecture-notes/code/lecture10/java/AStarGrid.java와 같은 정책(4-neighbor
 * unit-cost grid, Manhattan heuristic, stale-g-snapshot 검사, permanent
 * CLOSED, path reconstruction). */
public final class AStarGrid {
    public static final class Result {
        public List<int[]> path = new ArrayList<>();
        public int cost = Integer.MAX_VALUE;
        public int expanded;
        public int maxFrontier;
        public boolean goalDiscoveredBeforeExtraction;
        public boolean goalExtracted;
    }
    private static final class Entry implements Comparable<Entry> {
        final int r, c, g, f;
        Entry(int r, int c, int g, int f) { this.r=r; this.c=c; this.g=g; this.f=f; }
        public int compareTo(Entry o) {
            int x=Integer.compare(f,o.f); if(x!=0)return x;
            x=Integer.compare(g,o.g); if(x!=0)return x;
            x=Integer.compare(r,o.r); return x!=0?x:Integer.compare(c,o.c);
        }
    }
    private static final int[][] DIR={{-1,0},{0,-1},{0,1},{1,0}};

    // snippet:a-star:start
    public Result search(boolean[][] blocked, int sr, int sc, int gr, int gc, boolean zeroHeuristic) {
        validate(blocked, sr, sc, gr, gc);
        int rows=blocked.length, cols=blocked[0].length;
        int[][] g=new int[rows][cols], pr=new int[rows][cols], pc=new int[rows][cols];
        boolean[][] closed=new boolean[rows][cols];
        for(int[] row:g)Arrays.fill(row,Integer.MAX_VALUE);
        for(int[] row:pr)Arrays.fill(row,-1);
        for(int[] row:pc)Arrays.fill(row,-1);
        PriorityQueue<Entry> open=new PriorityQueue<>();
        g[sr][sc]=0; open.add(new Entry(sr,sc,0,h(sr,sc,gr,gc,zeroHeuristic)));
        Result out=new Result();
        while(!open.isEmpty()){
            out.maxFrontier=Math.max(out.maxFrontier,open.size());
            Entry e=open.remove();
            if(e.g!=g[e.r][e.c])continue;
            if(closed[e.r][e.c])continue;
            closed[e.r][e.c]=true;
            out.expanded++;
            if(e.r==gr&&e.c==gc){
                out.goalExtracted=true;
                out.cost=e.g;
                break;
            }
            for(int[] d:DIR){
                int nr=e.r+d[0],nc=e.c+d[1];
                if(nr<0||nr>=rows||nc<0||nc>=cols||blocked[nr][nc])continue;
                if(closed[nr][nc])continue;
                if(e.g==Integer.MAX_VALUE)continue;
                int ng=e.g+1;
                if(ng<g[nr][nc]){
                    g[nr][nc]=ng;pr[nr][nc]=e.r;pc[nr][nc]=e.c;
                    if(nr==gr&&nc==gc)out.goalDiscoveredBeforeExtraction=true;
                    open.add(new Entry(nr,nc,ng,ng+h(nr,nc,gr,gc,zeroHeuristic)));
                }
            }
        }
        if(out.cost!=Integer.MAX_VALUE){
            int r=gr,c=gc;
            while(r!=-1){out.path.add(new int[]{r,c});int nr=pr[r][c];c=pc[r][c];r=nr;}
            Collections.reverse(out.path);
        }
        return out;
    }
    private static int h(int r,int c,int gr,int gc,boolean zero){
        return zero?0:Math.abs(r-gr)+Math.abs(c-gc);
    }
    // snippet:a-star:end
    private static void validate(boolean[][] b,int sr,int sc,int gr,int gc){
        if(b==null||b.length==0||b[0].length==0)throw new IllegalArgumentException();
        int cols=b[0].length;for(boolean[] row:b)if(row==null||row.length!=cols)throw new IllegalArgumentException();
        if(sr<0||sr>=b.length||gr<0||gr>=b.length||sc<0||sc>=cols||gc<0||gc>=cols||b[sr][sc]||b[gr][gc])throw new IllegalArgumentException();
    }
}
